File: Fanficly/Search/SearchPromptParser.swift
import Foundation

public struct SearchPromptParser: Sendable {
    public init() {}

    public func parse(_ prompt: String) -> AO3SearchFilters {
        var filters = AO3SearchFilters()
        
        let normalizedPrompt = prompt
            .replacingOccurrences(of: "“", with: "\"")
            .replacingOccurrences(of: "”", with: "\"")
            .replacingOccurrences(of: "‘", with: "'")
            .replacingOccurrences(of: "’", with: "'")
            
        let working = NSMutableString(string: normalizedPrompt.lowercased())
        let casedWorking = NSMutableString(string: normalizedPrompt)

        extractTitle(from: working, casedWorking: casedWorking, into: &filters)
        extractCreators(from: working, casedWorking: casedWorking, into: &filters)
        extractExclusions(from: working, into: &filters)
        extractQuoted(from: working, into: &filters)
        extractShips(from: working, into: &filters)
        extractWordCount(from: working, into: &filters)
        extractEngagement(from: working, into: &filters)
        extractStatus(from: working, into: &filters)
        extractCrossover(from: working, into: &filters)
        extractLanguage(from: working, into: &filters)
        extractCategories(from: working, into: &filters)
        extractRatings(from: working, into: &filters)
        extractWarnings(from: working, into: &filters)
        extractFandoms(from: working, into: &filters)
        extractFreeformTags(from: working, into: &filters)

        let remainder = collapseWhitespace(working as String)
        if !remainder.isEmpty {
            filters.query = remainder
        }
        return filters
    }

    private func extractTitle(from working: NSMutableString, casedWorking: NSMutableString, into filters: inout AO3SearchFilters) {
        let patterns = [
            #"title:\s*"([^"]+)""#,
            #"title:\s*([^\s]+)"#
        ]
        for pattern in patterns {
            guard let regex = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive) else { continue }
            let matches = regex.matches(in: working as String, range: NSRange(location: 0, length: working.length))
            for match in matches.reversed() where match.numberOfRanges > 1 {
                let nsRange = match.range(at: 1)
                let extractedTitle = casedWorking.substring(with: nsRange).trimmingCharacters(in: .whitespaces)
                if !extractedTitle.isEmpty {
                    filters.title = extractedTitle
                }
                working.deleteCharacters(in: match.range)
                casedWorking.deleteCharacters(in: match.range)
            }
        }
    }

    private func extractCreators(from working: NSMutableString, casedWorking: NSMutableString, into filters: inout AO3SearchFilters) {
        let patterns = [
            #"\b(?:author|creator|creators|by):\s*"([^"]+)""#,
            #"\b(?:author|creator|creators|by):\s*([^\s]+)"#
        ]
        for pattern in patterns {
            guard let regex = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive) else { continue }
            let matches = regex.matches(in: working as String, range: NSRange(location: 0, length: working.length))
            for match in matches.reversed() where match.numberOfRanges > 1 {
                let nsRange = match.range(at: 1)
                let extractedCreators = casedWorking.substring(with: nsRange).trimmingCharacters(in: .whitespaces)
                if !extractedCreators.isEmpty {
                    filters.creators = extractedCreators
                }
                working.deleteCharacters(in: match.range)
                casedWorking.deleteCharacters(in: match.range)
            }
        }
    }

    private func extractExclusions(from working: NSMutableString, into filters: inout AO3SearchFilters) {
        let patterns = [
            #"(?:^|\s)-"([^"]+)""#,
            #"(?:^|\s)-([a-z0-9][a-z0-9'\-]*)\b"#,
            #"(?:^|\s)not\s+"([^"]+)""#,
            #"(?:^|\s)not\s+([a-z0-9][a-z0-9'\-]*)\b"#,
        ]
        for pattern in patterns {
            guard let regex = try? NSRegularExpression(pattern: pattern) else { continue }
            let matches = regex.matches(in: working as String, range: NSRange(location: 0, length: working.length))
            for match in matches.reversed() where match.numberOfRanges > 1 {
                let term = (working.substring(with: match.range(at: 1)))
                    .trimmingCharacters(in: .whitespaces)
                if !term.isEmpty {
                    filters.excludedFreeforms.append(term)
                }
                working.deleteCharacters(in: match.range)
            }
        }
    }

    /// Pulls out double-quoted phrases as exact tags before the ship regex can
    /// mangle them — so `"Hermione Granger/Draco Malfoy"` stays one relationship
    /// instead of the ship matcher grabbing "granger/draco". A quoted phrase with
    /// a slash is a relationship; otherwise it's a freeform tag. (Exclusions like
    /// `-"…"` were already consumed above.) Resolution later canonicalizes it.
    private func extractQuoted(from working: NSMutableString, into filters: inout AO3SearchFilters) {
        guard let regex = try? NSRegularExpression(pattern: #""([^"]+)""#) else { return }
        let matches = regex.matches(in: working as String, range: NSRange(location: 0, length: working.length))
        for match in matches.reversed() where match.numberOfRanges > 1 {
            let phrase = working.substring(with: match.range(at: 1)).trimmingCharacters(in: .whitespaces)
            if !phrase.isEmpty {
                if phrase.contains("/") {
                    filters.relationshipNames.append(phrase.capitalized)
                } else {
                    filters.freeformNames.append(phrase.capitalized)
                }
            }
            working.deleteCharacters(in: match.range)
        }
    }

    private func extractShips(from working: NSMutableString, into filters: inout AO3SearchFilters) {
        let pattern = #"(?<![\w/])([a-z][a-z'\-]+)\/([a-z][a-z'\-]+)(?![\w/])"#
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return }
        let matches = regex.matches(in: working as String, range: NSRange(location: 0, length: working.length))
        for match in matches.reversed() where match.numberOfRanges == 3 {
            let left = working.substring(with: match.range(at: 1))
            let right = working.substring(with: match.range(at: 2))
            if let category = KnownTags.categoryAliases["\(left)/\(right)"] {
                filters.categories.insert(category)
            } else {
                let ship = "\(left.capitalized)/\(right.capitalized)"
                filters.relationshipNames.append(ship)
            }
            working.deleteCharacters(in: match.range)
        }
    }

    private func extractWordCount(from working: NSMutableString, into filters: inout AO3SearchFilters) {
        struct Pattern {
            let regex: String
            let render: (Int, Int?) -> String
        }
        let kSuffix = #"(?:k|,000)"#
        let patterns: [Pattern] = [
            Pattern(regex: #"under\s+(\d+)\s*\#(kSuffix)?"#) { n, _ in "<\(n)" },
            Pattern(regex: #"over\s+(\d+)\s*\#(kSuffix)?"#) { n, _ in ">\(n)" },
            Pattern(regex: #"more than\s+(\d+)\s*\#(kSuffix)?"#) { n, _ in ">\(n)" },
            Pattern(regex: #"less than\s+(\d+)\s*\#(kSuffix)?"#) { n, _ in "<\(n)" },
            Pattern(regex: #"between\s+(\d+)\s*\#(kSuffix)?\s+and\s+(\d+)\s*\#(kSuffix)?"#) { lo, hi in
                if let hi { return "\(lo)-\(hi)" } else { return "\(lo)" }
            },
        ]
        for pattern in patterns {
            guard let regex = try? NSRegularExpression(pattern: pattern.regex) else { continue }
            let str = working as String
            let matches = regex.matches(in: str, range: NSRange(location: 0, length: working.length))
            for match in matches.reversed() {
                let raw1 = working.substring(with: match.range(at: 1))
                let full = working.substring(with: match.range)
                let hadK = full.contains("k") || full.contains(",000")
                let n1 = (Int(raw1) ?? 0) * (hadK ? 1_000 : 1)
                var n2: Int?
                if match.numberOfRanges > 2 {
                    let r2 = match.range(at: 2)
                    if r2.location != NSNotFound {
                        let raw2 = working.substring(with: r2)
                        let hadK2 = full.range(of: raw2 + "k") != nil || full.range(of: raw2 + ",000") != nil
                        n2 = (Int(raw2) ?? 0) * (hadK2 ? 1_000 : 1)
                    }
                }
                filters.wordCount = pattern.render(n1, n2)
                working.deleteCharacters(in: match.range)
            }
        }
    }

    private func extractEngagement(from working: NSMutableString, into filters: inout AO3SearchFilters) {
        if removeFirst(in: working, pattern: #"\bpopular\b"#) {
            filters.kudosCount = ">1000"
        }
        if removeFirst(in: working, pattern: #"\bmost kudos\b"#) {
            filters.sortColumn = .kudosCount
            filters.sortDirection = .desc
        }
        if removeFirst(in: working, pattern: #"\bmost hits\b"#) {
            filters.sortColumn = .hits
            filters.sortDirection = .desc
        }
        if removeFirst(in: working, pattern: #"\brecently updated\b"#) {
            filters.sortColumn = .revisedAt
            filters.sortDirection = .desc
        }
        if removeFirst(in: working, pattern: #"\bnewest\b|\brecent\b"#) {
            filters.sortColumn = .createdAt
            filters.sortDirection = .desc
        }
    }

    private func extractStatus(from working: NSMutableString, into filters: inout AO3SearchFilters) {
        if removeFirst(in: working, pattern: #"\boneshot\b|\bone\s*-?\s*shot\b|\bone shot\b"#) {
            filters.singleChapter = true
        }
        if removeFirst(in: working, pattern: #"\bcomplete\b|\bcompleted\b|\bfinished\b"#) {
            filters.complete = .yes
        }
        if removeFirst(in: working, pattern: #"\bwip\b|\bwork in progress\b|\bin progress\b|\bunfinished\b"#) {
            filters.complete = .no
        }
    }

    private func extractCrossover(from working: NSMutableString, into filters: inout AO3SearchFilters) {
        if removeFirst(in: working, pattern: #"\bno\s+crossovers?\b"#) {
            filters.crossover = .no
        } else if removeFirst(in: working, pattern: #"\bcrossovers?\b"#) {
            filters.crossover = .yes
        }
    }

    private func extractLanguage(from working: NSMutableString, into filters: inout AO3SearchFilters) {
        for (alias, code) in KnownTags.languageAliases {
            let escaped = NSRegularExpression.escapedPattern(for: alias)
            if removeFirst(in: working, pattern: #"\b\#(escaped)\b"#) {
                filters.languageId = code
                return
            }
        }
    }

    private func extractCategories(from working: NSMutableString, into filters: inout AO3SearchFilters) {
        for (alias, category) in KnownTags.categoryAliases {
            let escaped = NSRegularExpression.escapedPattern(for: alias)
            if removeFirst(in: working, pattern: #"(?<![\w/])\#(escaped)(?![\w/])"#) {
                filters.categories.insert(category)
            }
        }
    }

    private func extractRatings(from working: NSMutableString, into filters: inout AO3SearchFilters) {
        let sortedKeys = KnownTags.ratingAliases.keys.sorted { $0.count > $1.count }
        for alias in sortedKeys {
            guard let rating = KnownTags.ratingAliases[alias] else { continue }
            let escaped = NSRegularExpression.escapedPattern(for: alias)
            let pattern = alias.count == 1 ? #"\b\#(escaped)-rated\b"# : #"\b\#(escaped)\b"#
            if removeFirst(in: working, pattern: pattern) {
                filters.ratings.insert(rating)
            }
        }
    }

    private func extractWarnings(from working: NSMutableString, into filters: inout AO3SearchFilters) {
        let sortedKeys = KnownTags.warningAliases.keys.sorted { $0.count > $1.count }
        for alias in sortedKeys {
            guard let warning = KnownTags.warningAliases[alias] else { continue }
            let escaped = NSRegularExpression.escapedPattern(for: alias)
            if removeFirst(in: working, pattern: #"\b\#(escaped)\b"#) {
                filters.warnings.insert(warning)
            }
        }
    }

    private func extractFreeformTags(from working: NSMutableString, into filters: inout AO3SearchFilters) {
        let sortedTags = KnownTags.freeformTags.sorted { $0.count > $1.count }
        for tag in sortedTags {
            let escaped = NSRegularExpression.escapedPattern(for: tag)
            if removeFirst(in: working, pattern: #"\b\#(escaped)\b"#) {
                filters.freeformNames.append(canonicalize(tag))
            }
        }
    }

    private func extractFandoms(from working: NSMutableString, into filters: inout AO3SearchFilters) {
        let sortedAliases = KnownTags.fandomAliases.keys.sorted { $0.count > $1.count }
        for alias in sortedAliases {
            guard let canonical = KnownTags.fandomAliases[alias] else { continue }
            let escaped = NSRegularExpression.escapedPattern(for: alias)
            if removeFirst(in: working, pattern: #"\b\#(escaped)\b"#) {
                if !filters.fandomNames.contains(canonical) {
                    filters.fandomNames.append(canonical)
                }
            }
        }
    }

    private func canonicalize(_ tag: String) -> String {
        tag.split(separator: " ").map { word in
            let lower = word.lowercased()
            if lower == "of" || lower == "and" || lower == "the" || lower == "to" || lower == "a" {
                return lower
            }
            return word.prefix(1).uppercased() + word.dropFirst().lowercased()
        }.joined(separator: " ")
    }

    @discardableResult
    private func removeFirst(in working: NSMutableString, pattern: String) -> Bool {
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return false }
        let range = NSRange(location: 0, length: working.length)
        guard let match = regex.firstMatch(in: working as String, range: range) else { return false }
        working.deleteCharacters(in: match.range)
        return true
    }

    private func collapseWhitespace(_ s: String) -> String {
        s.replacingOccurrences(of: #"\s+"#, with: " ", options: .regularExpression)
            .trimmingCharacters(in: .whitespaces)
    }
}

File: android/app/src/main/java/io/github/yennster/fanficly/search/SearchPromptParser.kt
package io.github.yennster.fanficly.search

import io.github.yennster.fanficly.model.AO3SearchFilters

/**
 * Rules-based natural-language search parser — the Kotlin port of the iOS
 * `SearchPromptParser`. Turns a free-text prompt like
 * `edward/bella romance all human complete` into a structured
 * [AO3SearchFilters]. Each extractor consumes the substring it recognizes from
 * a mutable working copy of the prompt; whatever is left over becomes the
 * free-text `query`.
 *
 * The logic mirrors the iOS implementation one-to-one, including extractor
 * order (which is significant — e.g. word-count and status are pulled before
 * the leftover words could be mistaken for freeform tags). Keep the two in
 * sync; the shared `SearchPromptParserTests` cases are ported verbatim.
 */
class SearchPromptParser {

    fun parse(prompt: String): AO3SearchFilters {
        val filters = AO3SearchFilters()

        val normalized = prompt
            .replace('“', '"').replace('”', '"')   // “ ” → "
            .replace('‘', '\'').replace('’', '\'') // ‘ ’ → '

        val working = StringBuilder(normalized.lowercase())
        val cased = StringBuilder(normalized)

        extractTitle(working, cased, filters)
        extractCreators(working, cased, filters)
        extractExclusions(working, filters)
        extractQuoted(working, filters)
        extractShips(working, filters)
        extractWordCount(working, filters)
        extractEngagement(working, filters)
        extractStatus(working, filters)
        extractCrossover(working, filters)
        extractLanguage(working, filters)
        extractCategories(working, filters)
        extractRatings(working, filters)
        extractWarnings(working, filters)
        extractFandoms(working, filters)
        extractFreeformTags(working, filters)

        val remainder = collapseWhitespace(working.toString())
        if (remainder.isNotEmpty()) filters.query = remainder
        return filters
    }

    // MARK: - Field extractors (title / creators read the case-preserved copy)

    private fun extractTitle(working: StringBuilder, cased: StringBuilder, filters: AO3SearchFilters) {
        val patterns = listOf(
            "title:\\s*\"([^\"]+)\"",
            "title:\\s*([^\\s]+)",
        )
        extractField(working, cased, patterns) { filters.title = it }
    }

    private fun extractCreators(working: StringBuilder, cased: StringBuilder, filters: AO3SearchFilters) {
        val patterns = listOf(
            "\\b(?:author|creator|creators|by):\\s*\"([^\"]+)\"",
            "\\b(?:author|creator|creators|by):\\s*([^\\s]+)",
        )
        extractField(working, cased, patterns) { filters.creators = it }
    }

    /** Reads group 1 from the case-preserved copy, deletes the full match from both. */
    private fun extractField(
        working: StringBuilder,
        cased: StringBuilder,
        patterns: List<String>,
        assign: (String) -> Unit,
    ) {
        for (pattern in patterns) {
            val regex = Regex(pattern, RegexOption.IGNORE_CASE)
            for (m in regex.findAll(working.toString()).toList().asReversed()) {
                val g = m.groups[1]
                if (g != null) {
                    val value = cased.substring(g.range.first, g.range.last + 1).trim()
                    if (value.isNotEmpty()) assign(value)
                }
                working.delete(m.range.first, m.range.last + 1)
                cased.delete(m.range.first, m.range.last + 1)
            }
        }
    }

    private fun extractExclusions(working: StringBuilder, filters: AO3SearchFilters) {
        val patterns = listOf(
            "(?:^|\\s)-\"([^\"]+)\"",
            """(?:^|\s)-([a-z0-9][a-z0-9'\-]*)\b""",
            "(?:^|\\s)not\\s+\"([^\"]+)\"",
            """(?:^|\s)not\s+([a-z0-9][a-z0-9'\-]*)\b""",
        )
        for (pattern in patterns) {
            val regex = Regex(pattern)
            for (m in regex.findAll(working.toString()).toList().asReversed()) {
                val term = m.groupValues[1].trim()
                if (term.isNotEmpty()) filters.excludedFreeforms = filters.excludedFreeforms + term
                working.delete(m.range.first, m.range.last + 1)
            }
        }
    }

    /**
     * Pulls out double-quoted phrases as exact tags before the ship regex can
     * mangle them — so `"Hermione Granger/Draco Malfoy"` stays one relationship
     * instead of the ship matcher grabbing "granger/draco". A quoted phrase with
     * a slash is a relationship; otherwise it's a freeform tag.
     */
    private fun extractQuoted(working: StringBuilder, filters: AO3SearchFilters) {
        val regex = Regex("\"([^\"]+)\"")
        for (m in regex.findAll(working.toString()).toList().asReversed()) {
            val phrase = m.groupValues[1].trim()
            if (phrase.isNotEmpty()) {
                if (phrase.contains("/")) {
                    filters.relationshipNames = filters.relationshipNames + capitalizedWords(phrase)
                } else {
                    filters.freeformNames = filters.freeformNames + capitalizedWords(phrase)
                }
            }
            working.delete(m.range.first, m.range.last + 1)
        }
    }

    private fun extractShips(working: StringBuilder, filters: AO3SearchFilters) {
        val regex = Regex("""(?<![\w/])([a-z][a-z'\-]+)/([a-z][a-z'\-]+)(?![\w/])""")
        for (m in regex.findAll(working.toString()).toList().asReversed()) {
            val left = m.groupValues[1]
            val right = m.groupValues[2]
            val category = KnownTags.categoryAliases["$left/$right"]
            if (category != null) {
                filters.categories = filters.categories + category
            } else {
                filters.relationshipNames =
                    filters.relationshipNames + "${capitalizedWords(left)}/${capitalizedWords(right)}"
            }
            working.delete(m.range.first, m.range.last + 1)
        }
    }

    private class WordCountPattern(val regex: Regex, val render: (Int, Int?) -> String)

    private val wordCountPatterns: List<WordCountPattern> = run {
        val k = "(?:k|,000)"
        listOf(
            WordCountPattern(Regex("""under\s+(\d+)\s*$k?""")) { n, _ -> "<$n" },
            WordCountPattern(Regex("""over\s+(\d+)\s*$k?""")) { n, _ -> ">$n" },
            WordCountPattern(Regex("""more than\s+(\d+)\s*$k?""")) { n, _ -> ">$n" },
            WordCountPattern(Regex("""less than\s+(\d+)\s*$k?""")) { n, _ -> "<$n" },
            WordCountPattern(Regex("""between\s+(\d+)\s*$k?\s+and\s+(\d+)\s*$k?""")) { lo, hi ->
                if (hi != null) "$lo-$hi" else "$lo"
            },
        )
    }

    private fun extractWordCount(working: StringBuilder, filters: AO3SearchFilters) {
        for (pattern in wordCountPatterns) {
            for (m in pattern.regex.findAll(working.toString()).toList().asReversed()) {
                val full = m.value
                val raw1 = m.groupValues[1]
                val hadK = full.contains("k") || full.contains(",000")
                val n1 = (raw1.toIntOrNull() ?: 0) * (if (hadK) 1_000 else 1)
                var n2: Int? = null
                if (m.groups.size > 2 && m.groups[2] != null) {
                    val raw2 = m.groupValues[2]
                    val hadK2 = full.contains(raw2 + "k") || full.contains(raw2 + ",000")
                    n2 = (raw2.toIntOrNull() ?: 0) * (if (hadK2) 1_000 else 1)
                }
                filters.wordCount = pattern.render(n1, n2)
                working.delete(m.range.first, m.range.last + 1)
            }
        }
    }

    private fun extractEngagement(working: StringBuilder, filters: AO3SearchFilters) {
        if (removeFirst(working, """\bpopular\b""")) filters.kudosCount = ">1000"
        if (removeFirst(working, """\bmost kudos\b""")) {
            filters.sortColumn = AO3SearchFilters.SortColumn.KUDOS_COUNT
            filters.sortDirection = AO3SearchFilters.SortDirection.DESC
        }
        if (removeFirst(working, """\bmost hits\b""")) {
            filters.sortColumn = AO3SearchFilters.SortColumn.HITS
            filters.sortDirection = AO3SearchFilters.SortDirection.DESC
        }
        if (removeFirst(working, """\brecently updated\b""")) {
            filters.sortColumn = AO3SearchFilters.SortColumn.REVISED_AT
            filters.sortDirection = AO3SearchFilters.SortDirection.DESC
        }
        if (removeFirst(working, """\bnewest\b|\brecent\b""")) {
            filters.sortColumn = AO3SearchFilters.SortColumn.CREATED_AT
            filters.sortDirection = AO3SearchFilters.SortDirection.DESC
        }
    }

    private fun extractStatus(working: StringBuilder, filters: AO3SearchFilters) {
        if (removeFirst(working, """\boneshot\b|\bone\s*-?\s*shot\b|\bone shot\b""")) filters.singleChapter = true
        if (removeFirst(working, """\bcomplete\b|\bcompleted\b|\bfinished\b""")) filters.complete = AO3SearchFilters.TriState.YES
        if (removeFirst(working, """\bwip\b|\bwork in progress\b|\bin progress\b|\bunfinished\b""")) filters.complete = AO3SearchFilters.TriState.NO
    }

    private fun extractCrossover(working: StringBuilder, filters: AO3SearchFilters) {
        if (removeFirst(working, """\bno\s+crossovers?\b""")) {
            filters.crossover = AO3SearchFilters.TriState.NO
        } else if (removeFirst(working, """\bcrossovers?\b""")) {
            filters.crossover = AO3SearchFilters.TriState.YES
        }
    }

    private fun extractLanguage(working: StringBuilder, filters: AO3SearchFilters) {
        for ((alias, code) in KnownTags.languageAliases) {
            if (removeFirst(working, """\b${Regex.escape(alias)}\b""")) {
                filters.languageId = code
                return
            }
        }
    }

    private fun extractCategories(working: StringBuilder, filters: AO3SearchFilters) {
        for ((alias, category) in KnownTags.categoryAliases) {
            if (removeFirst(working, """(?<![\w/])${Regex.escape(alias)}(?![\w/])""")) {
                filters.categories = filters.categories + category
            }
        }
    }

    private fun extractRatings(working: StringBuilder, filters: AO3SearchFilters) {
        for (alias in KnownTags.ratingAliases.keys.sortedByDescending { it.length }) {
            val rating = KnownTags.ratingAliases[alias] ?: continue
            val escaped = Regex.escape(alias)
            val pattern = if (alias.length == 1) """\b$escaped-rated\b""" else """\b$escaped\b"""
            if (removeFirst(working, pattern)) filters.ratings = filters.ratings + rating
        }
    }

    private fun extractWarnings(working: StringBuilder, filters: AO3SearchFilters) {
        for (alias in KnownTags.warningAliases.keys.sortedByDescending { it.length }) {
            val warning = KnownTags.warningAliases[alias] ?: continue
            if (removeFirst(working, """\b${Regex.escape(alias)}\b""")) filters.warnings = filters.warnings + warning
        }
    }

    private fun extractFreeformTags(working: StringBuilder, filters: AO3SearchFilters) {
        for (tag in KnownTags.freeformTags.sortedByDescending { it.length }) {
            if (removeFirst(working, """\b${Regex.escape(tag)}\b""")) {
                filters.freeformNames = filters.freeformNames + canonicalize(tag)
            }
        }
    }

    private fun extractFandoms(working: StringBuilder, filters: AO3SearchFilters) {
        for (alias in KnownTags.fandomAliases.keys.sortedByDescending { it.length }) {
            val canonical = KnownTags.fandomAliases[alias] ?: continue
            if (removeFirst(working, """\b${Regex.escape(alias)}\b""")) {
                if (!filters.fandomNames.contains(canonical)) filters.fandomNames = filters.fandomNames + canonical
            }
        }
    }

    // MARK: - Helpers

    /** Title-cases a freeform tag while keeping short joining words lowercase. */
    private fun canonicalize(tag: String): String =
        tag.split(" ").joinToString(" ") { word ->
            val lower = word.lowercase()
            when {
                word.isEmpty() -> word
                lower == "of" || lower == "and" || lower == "the" || lower == "to" || lower == "a" -> lower
                else -> word.substring(0, 1).uppercase() + word.substring(1).lowercase()
            }
        }

    /**
     * Swift `String.capitalized` equivalent: capitalize the first letter of each
     * word and lowercase the rest, using Unicode word breaking. Spaces and `/`
     * start a new word, but a mid-word apostrophe (between two letters) does
     * NOT — so `coffee shop au` → `Coffee Shop Au`, `draco/hermione` →
     * `Draco/Hermione`, and `don't tell` → `Don't Tell` (not `Don'T Tell`),
     * matching the iOS parser exactly.
     */
    private fun capitalizedWords(s: String): String {
        val sb = StringBuilder(s.length)
        var atWordStart = true
        for (i in s.indices) {
            val ch = s[i]
            if (ch.isLetter()) {
                sb.append(if (atWordStart) ch.uppercaseChar() else ch.lowercaseChar())
                atWordStart = false
            } else {
                sb.append(ch)
                val midWordApostrophe = ch == '\'' &&
                    i > 0 && s[i - 1].isLetter() &&
                    i + 1 < s.length && s[i + 1].isLetter()
                if (!midWordApostrophe) atWordStart = true
            }
        }
        return sb.toString()
    }

    /** Deletes the first match of [pattern] from [working]; true if one was removed. */
    private fun removeFirst(working: StringBuilder, pattern: String): Boolean {
        val m = Regex(pattern).find(working) ?: return false
        working.delete(m.range.first, m.range.last + 1)
        return true
    }

    private fun collapseWhitespace(s: String): String =
        s.replace(Regex("""\s+"""), " ").trim()
}
